### PR TITLE
Windows code fixes

### PR DESCRIPTION
**Pull request by JohannesLorenz**
Please review, thanks on advance.

Suggesting to look at commits separately.

In case you merge, please use rebase-and-ff, not squash.

### CODE REVIEW

File: include/rtosc/bundle-foreach.h [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017 Johannes Lorenz
+ * Copyright (c) 2017-2023 Johannes Lorenz
```
Comment (fundamental) [inline include/rtosc/bundle-foreach.h:2]: minor non-blocker, but it is 2024 after all :p

File: include/rtosc/bundle-foreach.h [hunk 2]
```diff
@@ -64,7 +64,7 @@ void bundle_foreach(const struct Port& p, const char* name, char* old_end,
     for(unsigned i=0; i<max; ++i)
     {
         const char* name2_2 = name;
-        pos2 = pos + sprintf(pos,"%d",i);
+        pos2 = pos + snprintf(pos,16,"%d",i);  // TODO: check length is >= 16
```
Comment (fundamental) [inline include/rtosc/bundle-foreach.h:67]: This feels cheaty since this commit doesn't update the API of bundle_foreach to explicitly accept a true length.

File: src/cpp/ports.cpp [hunk 3]
```diff
@@ -1106,15 +1106,15 @@ static void walk_ports_recurse0(const Port& p, char* name_buffer,
         if(*read_head == '/') { ++read_head; }
         if(ranges)
         {
-            int written = sprintf(write_head,"[0,%d]/", max-1);
+            int written = snprintf(write_head,32,"[0,%d]/", max-1);  // TODO: make sure 32 fits
```
Comment (fundamental) [inline src/cpp/ports.cpp:1109]: Same bit of cheatyness since we don't know the true length.

File: src/cpp/automations.cpp [hunk 4]
```diff
@@ -12,7 +12,7 @@ AutomationMgr::AutomationMgr(int slots, int per_slot, int control_points)
     memset(this->slots, 0, sizeof(AutomationSlot)*slots);
     for(int i=0; i<slots; ++i) {
         auto &s = this->slots[i];
-        sprintf(s.name, "Slot %d", i + 1);
+        snprintf(s.name, sizeof(s.name), "Slot %d", i + 1);
```
Comment (fundamental) [inline src/cpp/automations.cpp:15]: Perfectly reasonable

File: src/cpp/default-value.cpp [hunk 5]
```diff
@@ -107,8 +107,12 @@ int get_default_value(const char* port_name, const char* port_args,
     if(pretty)
     {
         nargs = rtosc_count_printed_arg_vals(pretty);
+#ifdef NDEBUG
```
Comment (fundamental) [inline src/cpp/default-value.cpp:110]: I'd personally leave the `(void)variable;` lines in place unconditionally, but that's a minor note.

**Comment by fundamental** [comment]
- [71d9e95](https://github.com/fundamental/rtosc/pull/75/commits/71d9e9586ca42c4198bd7b1973558f4797a4507d) I'm not sure why this is needed, but it shouldn't negatively impact any platforms
- [f3be6dd](https://github.com/fundamental/rtosc/pull/75/commits/f3be6ddd4a76e9b944b96bc074cd2df0eed3b529) is this really about removing trailing zeros or about truncating float64 representations into the float32 representations that you might expect? I haven't directly played around with the output of %a, so I might be off base here.
- [1ca80c2](https://github.com/fundamental/rtosc/pull/75/commits/1ca80c22c7562d598e515fb0ec98f971538615d5) I'm not thrilled by adding in false lengths just to bypass warnings. This feels more likely to hide or create bugs than to fix them.
- [655fe97](https://github.com/fundamental/rtosc/pull/75/commits/655fe971280db6b1cba98b0a9fc464da93991fad) Approved
- [3c17fed](https://github.com/fundamental/rtosc/pull/75/commits/3c17fed39fde899bd3751f77bde7de0df746d6f3) Approved
- [985c7ff](https://github.com/fundamental/rtosc/pull/75/commits/985c7ffb68b2428791c200010c51b3b322e945e4) Approved
- [26b46e0](https://github.com/fundamental/rtosc/pull/75/commits/26b46e06f4e40524950fb49293baa08fd94ffca9) Approved, though I thought we already had this in play
- [d36a31f](https://github.com/fundamental/rtosc/pull/75/commits/d36a31f065c1d8d2c023b0360df170e8dff6ba85) Looks reasonable to me, Approved
- [285e0cb](https://github.com/fundamental/rtosc/pull/75/commits/285e0cbfb3fb2258bb2d72ab1366115e3aff045a) trivial approved
- [ae72871](https://github.com/fundamental/rtosc/pull/75/commits/ae728710173f771ac8f7b24fff83202f39e4d5bb) I've got a minor stylistic comment, but approved since it's not a blocker
- [cf693c7](https://github.com/fundamental/rtosc/pull/75/commits/cf693c743048fda2d642885d8a6d13df56809150) Looks like a reasonable error handling style, approved

In summary I'd say let's not have 1ca80c2 (maybe temporarily suppress warnings with a pragma) and maybe double check f3be6dd is stripping out extra output in the desired way.

**Comment by JohannesLorenz** [comment]
> * is this really about removing trailing zeros or about truncating float64 representations into the float32 representations that you might expect? I haven't directly played around with the output of %a, so I might be off base here.

This was really about truncating zeroes in the "%a" output. "%a" is the lossless float representation. The problem is that this representation looks different on Windows than on Linux/Mac - there were different numbers of trailing zeroes, leading to test failure. I don't think this is float64 related, since float is said to always be handled as 4 bytes (IEEE 754) by MSVC. It might be that they have some weird internal conversions inside `printf` - but whatever it is, we needed to make the output on all systems identical, so I just cut all trailing zeroes (and trailing points) on all systems.

**Comment by fundamental** [comment]
Sounds reasonable enough to me. I was considering %a might have a weird platform dependence and I think the C standard dictates that floats are upgraded to doubles via va_args, but I could be incorrect there. That only leaves my concerns about the snprintf commit

File: src/cpp/default-value.cpp (hunk 5 above)
Comment (JohannesLorenz) [reply src/cpp/default-value.cpp:110]: I thought about this, too, but with `NDEBUG`, we will get notified if `n` should get completely unused in the future, so I'll keep it.

**Comment by JohannesLorenz** [comment]
@fundamental `snprintf` issues should be fixed - please check.

Please note that if you should accept the PR, I will still reorder and partially squash the commits before merge.

File: include/rtosc/bundle-foreach.h [hunk 6]
```diff
@@ -31,53 +31,75 @@
...
+                    void* const data, void* const runtime, const F& ftor,
                     /* options */
-                    bool expand_bundles = true,
-                    bool cut_afterwards = true,
-                    bool ranges = false)
+                    const bool expand_bundles = true,
+                    const bool cut_afterwards = true,
+                    const bool ranges = false)
 {
+    size_t space_left = buffer_size - (size_t)(old_end - name_buffer);
```
Comment (fundamental) [inline include/rtosc/bundle-foreach.h:80]: minor, but I'd use ssize_t here to keep things signed in this computation in case there is a buggy old_end/name_buffer input (aka to handle buggy cases where space_left would be negative).

File: src/cpp/savefile.cpp [hunk 7]
```diff
@@ -264,7 +264,7 @@ std::string get_changed_values(const Ports& ports, void* runtime,
 
             auto refix_old_end = [&base](const Port* _p, char* _old_end)
             { // TODO: remove base capture
-                bundle_foreach(*_p, _p->name, _old_end, NULL,
+                bundle_foreach(*_p, _p->name, _old_end, NULL, 0,
```
Comment (fundamental) [inline src/cpp/savefile.cpp:267]: Isn't this going to trigger an assert (the `0,` is setting the buffersize to zero and any zero size buffer would trigger an assert, no?)

**Comment by fundamental** [comment]
Possibly one issue found with a buffersize of zero, but otherwise the code seems to look correct on a second pass.

File: include/rtosc/bundle-foreach.h (hunk 6 above)
Comment (JohannesLorenz) [reply include/rtosc/bundle-foreach.h]: Fixed by using `ssize_t`.

File: src/cpp/savefile.cpp (hunk 7 above)
Comment (JohannesLorenz) [reply src/cpp/savefile.cpp]: That was indeed a problem (which I did not see because I forgot to run the tests, and the CI is not in place *yet*). Previously, in this context, the function did not even access the buffer and the buffer size, but it now does, for the size computations, so I fixed this function call.

**Comment by JohannesLorenz** [comment]
Should be all fixed now. Please do a quick rework check. Thanks on advance.

**Comment by fundamental** [comment]
All changes look good and are approved from my end.
